# Лабораторная работа №2
## Выполнил студент группы БВТ1901 Рыбаков Егор
### Задание №1
<p>Реализовать  методы  поиска  в  соответствии  с  заданием.  Организовать  генерацию 
начального  набора  случайных  данных.  Для  всех  вариантов  добавить  реализацию 
добавления,  поиска  и  удаления  элементов.  Оценить  время  работы  каждого  алгоритма 
поиска и сравнить его со временем работы стандартной функции поиска, используемой в 
выбранном языке программирования. </p>


#### 1 Бинарный поиск

In [55]:
from random import randint
import time

start = time.time()
def binarySearch(arr, key):
    left = 0
    right = len(arr) - 1
    while left <= right:
        mid = (right + left) // 2
        if key > arr[mid]:
            left = mid + 1
        elif key < arr[mid]:
            right = mid - 1
        else:
            return mid
    return "No results"


n = 1000
key = int(input("Введите число для поиска: "))
arr = [randint(-500, 500) for i in range(n)]
arr = sorted(arr)
print(binarySearch(arr, key))
print(time.time() - start)


Введите число для поиска: 219
734
1.6065092086791992


#### 2 Бинарное дерево

In [56]:
from random import randint
import time

start = time.time()
class Tree:
    def __init__(self, val):
        self.left = None
        self.right = None
        self.val = val

    def insert(self, val):
        if self.val:
            if val < self.val:
                if self.left is None:
                    self.left = Tree(val)
                else:
                    self.left.insert(val)
            else:
                if self.right is None:
                    self.right = Tree(val)
                else:
                    self.right.insert(val)
        else:
            self.val = val

    def search(self, val):
        if self.val == val:
            print("Success!", val, "найдено!")
            return self
        else:
            if val < self.val:
                if self.left is not None:
                    return self.left.search(val)
                else:
                    print(val, "not found!")
            elif val > self.val:
                if self.right is not None:
                    return self.right.search(val)
                else:
                    print(val, "not found!")

    def delete(self, obj):
        if self.left.val == obj.val:
            self.left = None
            return
        elif self.right.val == obj.val:
            self.right = None
            return
        return self.delete(obj.left)

    def insertLostValues(self, arr):
        for x in arr:
            Tree.insert(self, x)

    def collectValues(self, obj):
        arr = []
        if obj is not None:
            arr = self.collectValues(obj.left)
            obj.left = None
            arr.append(obj.val)
            arr += self.collectValues(obj.right)
            obj.right = None
        return arr

    def printTree(self):
        if self is not None:
            if self.left is not None:
                print("Root: ", self.val)
                print("L:", self.left.val)
                self.left.printTree()
            if self.right is not None:
                print("Root: ", self.val)
                print("R:", self.right.val)
                self.right.printTree()


n = 1000
key = int(input("Введите число для поиска: "))
arr = [randint(-500, 500) for i in range(n)]
root = Tree(arr[0])
for x in arr[1::]:
    root.insert(x)
root.search(key)
print(time.time() - start)

Введите число для поиска: 219
219 not found!
1.506004810333252


#### 3 Фибоначчиев поиск

In [9]:
from random import randint
import time

start = time.time()
def FibonacciSearch(arr, val):
    fibM_minus_2 = 0
    fibM_minus_1 = 1
    fibM = fibM_minus_1 + fibM_minus_2
    while fibM < len(arr):
        fibM_minus_2 = fibM_minus_1
        fibM_minus_1 = fibM
        fibM = fibM_minus_1 + fibM_minus_2
    index = -1;
    while fibM > 1:
        i = min(index + fibM_minus_2, (len(arr)-1))
        if arr[i] < val:
            fibM = fibM_minus_1
            fibM_minus_1 = fibM_minus_2
            fibM_minus_2 = fibM - fibM_minus_1
            index = i
        elif arr[i] > val:
            fibM = fibM_minus_2
            fibM_minus_1 = fibM_minus_1 - fibM_minus_2
            fibM_minus_2 = fibM - fibM_minus_1
        else :
            return i
    if(fibM_minus_1 and index < (len(arr)-1) and arr[index+1] == val):
        return index+1;
    return -1


key = int(input("Введите элемент для поиска: "))
n = 1000
arr = [randint(-500, 500) for i in range(n)]
arr = sorted(arr)
print(FibonacciSearch(arr, key))
print(time.time() - start)

Введите элемент для поиска: 459
954
1.683211088180542


#### 4 Интерполяционный поиск

In [12]:
from random import randint
import time


start = time.time()
def interSearch(arr, key):
    left = 0
    right = n - 1
    while arr[left] <= key <= arr[right]:
        mid = int(left + (((key - arr[left]) * (right - left)) / (arr[right] - arr[left])))
        if arr[mid] < key:
            left = mid + 1
        elif arr[mid] > key:
            right = mid - 1
        else:
            return mid
    return -1


n = 1000
arr = [randint(-500, 500) for i in range(n)]
arr = sorted(arr)
key = int(input("Введите элемент для поиска: "))
print(interSearch(arr, key))
print(time.time() - start)

Введите элемент для поиска: 183
692
1.3307123184204102


### Реализовать  методы хеширования данных.

#### 1 Простое рехэширование

In [43]:
from random import randint
import time


start = time.time()
class HashTable:
    def __init__(self, size):
        self.size = size + 1
        self.slots = [None] * self.size
        self.data = [None] * self.size

    def hashFunc(self, val):
        return val % self.size

    def addEl(self, val, k = 0):
        global savedVal
        if k == 0:
            savedVal = val
            k += 1
        hashVal = self.hashFunc(val)
        if self.slots[hashVal] is None:
            self.slots[hashVal] = hashVal
            self.data[hashVal] = savedVal
        else:
            self.addEl(hashVal + 1, k)

    def getElbyHash(self, val):
        return self.data[val]

    def getPositionByValue(self, val, k = 0):
        global savedVal1
        if k == 0:
            savedVal1 = val
            k += 1
        hashVal = self.hashFunc(val)
        if self.slots[hashVal] is not None:
            if self.data[hashVal] == savedVal1:
                return hashVal
            else:
                return self.getPositionByValue(hashVal + 1, k)
        else:
            return "No Results"

    def delElByHash(self, val):
        self.slots[val] = None
        self.data[val] = None

    def delElByValue(self, val):
        j = 0
        for i in range(len(self.data)):
            if self.data[i] == val:
                self.data[i] = None
                self.slots[i] = None
                j = 1
            if j == 1:
                for k in range(i, len(self.data) - 1):
                    self.data[k] = self.data[k + 1]
                    self.slots[k] = self.slots[k + 1]
                break


hashtable = HashTable(1000)
arr = [randint(-500, 500) for i in range(1000)]
for x in arr:
    hashtable.addEl(x)
print(hashtable.getPositionByValue(119))
print(time.time() - start)

125
0.026003599166870117


#### 2 Рехэширование с помощью псевдослучайных чисел

In [46]:
from random import choice, randint
import time


start = time.time()
class HashTable:
    def __init__(self, size):
        self.arr = [i for i in range(1, 1000)]
        self.size = size + 1
        self.slots = [None] * self.size
        self.data = [None] * self.size

    def hashFunc(self, val):
        return (val + choice(self.arr)) % self.size

    def addEl(self, val, k = 0):
        global savedVal
        if k == 0:
            savedVal = val
            k += 1
        hashVal = self.hashFunc(val)
        if self.slots[hashVal] is None:
            self.slots[hashVal] = hashVal
            self.data[hashVal] = savedVal
        else:
            self.addEl(hashVal + 1, k)

    def getElbyHash(self, val):
        return self.data[val]

    def getPositionByValue(self, val, k = 0):
        global savedVal1
        if k == 0:
            savedVal1 = val
        k += 1
        if k > 997:
            return "No results"
        else:
            hashVal = self.hashFunc(val)
            if self.slots[hashVal] is not None:
                if self.data[hashVal] == savedVal1:
                    return hashVal
                else:
                    return self.getPositionByValue(hashVal + 1, k)
            else:
                return "No Results"

    def delElByHash(self, val):
        self.slots[val] = None
        self.data[val] = None

    def delElByValue(self, val):
        for i in range(len(self.data)):
            if self.data[i] == val:
                self.data[i] = None
                self.slots[i] = None



hashtable = HashTable(1000)
arr = [randint(-500, 500) for i in range(1000)]
for x in arr:
    hashtable.addEl(x)
print(hashtable.getPositionByValue(119))
print(time.time() - start)

86
0.03900790214538574


#### 3 Метод цепочек
<p>Метод цепочек реализуется двумя файлами <br>
1 файл (Реализация связного списка):</p>

In [6]:
class cell:
    def __init__(self, val):
        self.val = val
        self.next = None


class LinkedList:
    def __init__(self):
        self.first = None
        self.last = None
        self.size = 0

    def add(self, val):
        if self.first is None:
            self.first = self.last = cell(val)
            self.size += 1
        else:
            self.last.next = self.last = cell(val)

    def deleteEl(self, val):
        if self.first.val == val:
            self.first = self.first.next
        else:
            current = self.first
            while current.next.val != val:
                current = current.next
            if current.next != self.last:
                current.next = current.next.next
            else:
                current.next = None

    def searchEl(self, val):
        current = self.first
        if current is not None:
            if current.val == val:
                return "{} is on 0 position".format(val)
            else:
                k = 1
                while current.next.val != val:
                    current = current.next
                    k += 1
                if current.next.val == val:
                    return "{} is on {} position".format(val, k)
                else:
                    return "value is not found"
        else:
            return "value is not found"

    def printList(self):
        current = self.first
        while current is not None:
            print(current.val)
            current = current.next

2 файл (Реализация хэширования, с помощью метода цепочек):

In [ ]:
import LinkedList
from random import randint


class ChainHash:
    def __init__(self, size):
        self.size = size
        self.slots = [None] * self.size
        self.data = [LinkedList.LinkedList() for x in range(self.size)]


    def hashFunc(self, val):
        return val % self.size

    def addEl(self, val):
        hashval = self.hashFunc(val)
        self.slots[hashval] = hashval
        self.data[hashval].add(val)

    def delEl(self, val):
        hashvalue = self.hashFunc(val)
        self.data[hashvalue].deleteEl(val)

    def search(self, val):
        hashval = self.hashFunc(val)
        ans = self.data[hashval].searchEl(val)
        return ans + "with hash - {}".format(hashval)


    def __str__(self):
        s = ""
        k = 0
        for x in self.slots:
            if x is not None:
                current = self.data[x].first
                s += "hash - " + str(x) + ": "
                while current is not None:
                    s += str(current.val) + " "
                    current = current.next
                s += '\n'
            else:
                s += "hash - " + str(k) + ": None"
                s += '\n'
            k += 1
        return s


chainhash = ChainHash(1000)
arr = [randint(-100000, 100000) for i in range(100000)]
for x in arr:
    chainhash.addEl(x)
print(chainhash.search(11119))

### Задание №3

In [53]:
import time

start = time.time()
board = [[0 for j in range(8)] for i in range(8)]
def setQueen(i, j):
    for x in range(8):
        board[x][j] += 1
        board[i][x] += 1
        foo = j - i + x
        if 0 <= foo < 8:
            board[x][foo] += 1
        foo = j + i - x
        if 0 <= foo < 8:
            board[x][foo] += 1
    board[i][j] = -1


def resetQueen(i, j):
    for x in range(8):
        board[x][j] -= 1
        board[i][x] -= 1
        foo = j - i + x
        if 0 <= foo < 8:
            board[x][foo] -= 1
        foo = j + i - x
        if 0 <= foo < 8:
            board[x][foo] -= 1
    board[i][j] = 0


def tryQueen(i):
    result = False
    for j in range(8):
        if board[i][j] == 0:
            setQueen(i, j)
            if i == 7:
                result = True
            else:
                result = tryQueen(i + 1)
                if not result:
                    resetQueen(i, j)
        if result:
            break
    return result


tryQueen(0)
for i in range(8):
    for j in range(8):
        if board[i][j] == -1:
            print("1", end="")
        else:
            print("0", end = "")
    print()

finish = time.time()
print()
print(finish - start)

10000000
00001000
00000001
00000100
00100000
00000010
01000000
00010000

0.007999658584594727


### Вывод
#### Search Algorithms:
<ul>
    <li>Binary Search        - 1.6065092086791992 </li>
    <li>Binary Tree          - 1.506004810333252 </li>
    <li>Fibonacci Search     - 1.683211088180542 </li>
    <li>Interpolation Search - 1.3307123184204102 </li>
</ul>

#### Hash Algorithms:

<ul>
    <li> Simple Rehash - 0.026003599166870117 </li>
    <li> Random Rehash - 0.03900790214538574 </li>
    <li> Chain Method - 0.08000445365905762 (С большим числом элементов, чем у прошлых двух алгоритмов) </li>
</ul>

#### Итог:

<p>Хэш-методы работают быстрее, чем методы поиска, т.к для их реализации данные могут быть не отсортированны.</p>